rust/gtk-client/src/main.rs

```diff
@@ -0,0 +1,384 @@
+const LOGO_BASE: &[u8] = include_bytes!("../../gui-client/src-tauri/icons/tray/Logo.png");
+const LOGO_GREY_BASE: &[u8] = include_bytes!("../../gui-client/src-tauri/icons/tray/Logo grey.png");
+const BUSY_LAYER: &[u8] = include_bytes!("../../gui-client/src-tauri/icons/tray/Busy layer.png");
+const SIGNED_OUT_LAYER: &[u8] =
+    include_bytes!("../../gui-client/src-tauri/icons/tray/Signed out layer.png");
+const UPDATE_READY_LAYER: &[u8] =
+    include_bytes!("../../gui-client/src-tauri/icons/tray/Update ready layer.png");
+
+const TOOLTIP: &str = "Firezone";
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct Cli {
+    #[command(subcommand)]
+    command: Option<Cmd>,
+}
+
+#[derive(clap::Subcommand)]
+enum Cmd {
+    Debug,
+    OpenDeepLink(DeepLink),
+}
+
+#[derive(Args)]
+struct DeepLink {
+    url: url::Url, // TODO: Should be `Secret`?
+}
+
+fn main() -> Result<()> {
+    let current_exe = std::env::current_exe()?;
+    let cli = Cli::parse();
+
+    match cli.command {
+        Some(Cmd::Debug) => return Ok(()), // I didn't want to use `if-let` here
+        Some(Cmd::OpenDeepLink(deep_link)) => {
+            let rt = tokio::runtime::Runtime::new()?;
+            if let Err(error) = rt.block_on(deep_link::open(&deep_link.url)) {
+                tracing::error!(?error, "Error in `OpenDeepLink`");
+            }
+            return Ok(());
+        }
+        None => {}
+    }
+
+    // We're not a deep link handler, so start telemetry
+    let telemetry = telemetry::Telemetry::default();
+    // TODO: Fix missing stuff for telemetry
+    telemetry.start(
+        "wss://api.firez.one",
+        firezone_bin_shared::git_version!("gtk-client-*"),
+        telemetry::GUI_DSN,
+    );
+
+    let common::logging::Handles {
+        logger: _logger,
+        reloader: log_filter_reloader,
+    } = start_logging("debug")?; // TODO
+    let rt = tokio::runtime::Runtime::new()?;
+    let _guard = rt.enter();
+
+    let deep_link_server = rt.block_on(async { deep_link::Server::new().await })?;
+
+    let app = Application::builder()
+        .application_id("dev.firezone.client")
+        .build();
+
+    app.connect_activate(|app| {
+        // We create the main window.
+        let win = ApplicationWindow::builder()
+            .application(app)
+            .default_width(640)
+            .default_height(480)
+            .title("Firezone GTK+ 3")
+            .build();
+
+        // Don't forget to make all widgets visible.
+        win.show_all();
+    });
+
+    gtk::init()?;
+    let tray_icon = TrayIconBuilder::new()
+        .with_tooltip(TOOLTIP)
+        .with_icon(icon_to_native_icon(&Icon::default()))
+        .build()?;
+
+    let (main_tx, main_rx) = mpsc::channel(100);
+
+    let l = MainThreadLoop {
+        app: app.clone(),
+        last_icon_set: Default::default(),
+        main_rx,
+        tray_icon,
+    };
+    glib::spawn_future_local(l.run());
```

Comment:
Where does the return-value of this future go? I think it is currently being dropped silently, right? I think we should only spawn a `-> ()` future here.

Reply:
Yeah, I'm guessing that future is only polled while `app.run` is running. I tried to join it afterwards and it just appeared to deadlock the main thread when I quit the app.

So I changed it so that function is not expected to return, and we log errors when handling main thread reqeusts